Made RKPathMatcher thread safe to fix #2133 and #1998

megakode opened the pull request:
RKPathMatcher had two instance variables sourcePath and rootPath, which was changed every time it was queried for a match. This was not thread safe and was the cause of two reported bugs. 

The common use-case where submitting restkit requests to a concurrent dispatch queue, can result in RKPathMatcher being called almost simultaneously, which is being messed up when sourcePath and rootPath is changed in the middle of everything.

I commented out matchesPattern method as it was not used and also messed with the socPattern variable. If we shall re-enable this method, we also have to remove the socPattern variable and pass that on a pr. method call basis.

An alternative fix to this is to never keep RKPathMatcher instances and create them on a pr. match basis in RKResponseDescriptor. I tried this first, and that also worked fine, but this was a better solution i think.

megakode commented:
The CI test fails because i commented out the unused matchesPattern:tokenizeQueryStrings:parsedArguments method btw.

segiddins commented:
I will take a look at this later, but I'll be unable to merge as long as tests are failing.

segiddins commented:
Can you just rewrite that method to use the new stateless methods?

megakode commented:
I can do one of two things:

1) remove the test that tries to call the method that i have now removed
2) Re-enable the unused method and re-write it to also be stateless. 

I would prefer solution 1, as the method changes the SOCPattern which PathMatcher is initialized with, which means that if SOCPattern should be able to change on an already initialized PathMatcher, we should pass it with every method call. Which i think is a bad solution. I would definitely prefer that at least this variable is fixed and not changed after initialisation.

Any preference on this?

megakode commented:
Ok, i've fixed it. Rewrote the internal methods a bit, so the class still have all the same public methods while being thread-safe. It comes at cost though: initiating a RKPathMatcher with a pattern can only match paths, and initiating a RKPathMatcher with a path can only match patterns. It was kinda like this always though, as stated in some of the NSAssert messages already in the class.

valeriomazzeo commented:
do you think this can be anyway related to this as well?

```
Thread : Crashed: com.apple.main-thread
0  libobjc.A.dylib                0x0000000197100174 objc_release + 20
1  MyProject                         0x000000010036bdf0 -[RKPathMatcher matchesPath:tokenizeQueryStrings:parsedArguments:] (RKPathMatcher.m:140)
2  MyProject                         0x00000001003724f4 -[RKResponseDescriptor matchesPath:] (RKResponseDescriptor.m:120)
3  MyProject                         0x000000010035c668 __RKFilteredArrayOfResponseDescriptorsMatchingPathAndMethod_block_invoke (RKObjectManager.m:73)
4  CoreFoundation                 0x0000000187692cd8 __NSArrayGetIndexesPassingTest + 576
5  MyProject                         0x0000000100358794 -[RKObjectManager appropriateObjectRequestOperationWithObject:method:path:parameters:] (RKObjectManager.m:72)
```

megakode commented:
@valeriomazzeo It's most likely this issue yes.

segiddins commented:
Excellent, @megakode! Is there any way you could add some test coverage?

bimusiek commented:
Hey, I am facing same problem. Using NSOperationQueue we are doing synchronistation with our backend and once in a while app crashes on `RKPatchMatcher.rootPath`, `malloc: double free`

+1 for a fix!

e5faf2 commented:
so, that about merge?

segiddins commented:
@e5faf5, as I said, this is waiting on added test coverage

megakode commented:
Sorry for the late reply, haven't really thought of a test for this as the existing functionality of the class is already covered by tests, and it's a thread issue which was pretty hard to replicate already.

bimusiek commented:
@megakode We have your fix already in production for almost one month and we got rid of all restkit crashes. So it was issue for this time. 
Thanks :)
